runtime/runtime/src/congestion_control.rs

```diff
@@ -0,0 +1,372 @@
+    fn buffer_receipt(
+        &mut self,
+        receipt: &Receipt,
+        state_update: &mut TrieUpdate,
+        shard: u64,
+        config: &RuntimeConfig,
+    ) -> Result<(), RuntimeError> {
+        let bytes = receipt_size(&receipt)?;
+        let gas = receipt_congestion_gas(&receipt, config)?;
+        self.congestion_info.add_receipt_bytes(bytes as u64)?;
+        self.congestion_info.add_buffered_receipt_gas(gas)?;
+        self.outgoing_buffers.to_shard(shard).push(state_update, &receipt)?;
+        Ok(())
+    }
+}
+
+fn receipt_congestion_gas(
+    receipt: &Receipt,
+    config: &RuntimeConfig,
+) -> Result<Gas, IntegerOverflowError> {
+    match &receipt.receipt {
+        ReceiptEnum::Action(action_receipt) => {
+            // account for gas guaranteed to be used for executing the receipts
+            let prepaid_exec_gas = safe_add_gas(
+                total_prepaid_exec_fees(config, &action_receipt.actions, &receipt.receiver_id)?,
+                config.fees.fee(ActionCosts::new_action_receipt).exec_fee(),
+            )?;
+            // account for gas guaranteed to be used for creating new receipts
+            let prepaid_send_gas = total_prepaid_send_fees(config, &action_receipt.actions)?;
+            let prepaid_gas = safe_add_gas(prepaid_exec_gas, prepaid_send_gas)?;
+
+            // account for gas potentially used for dynamic execution
+            let gas_attached_to_fns = total_prepaid_gas(&action_receipt.actions)?;
+            let gas = safe_add_gas(gas_attached_to_fns, prepaid_gas)?;
+
+            Ok(gas)
+        }
+        ReceiptEnum::Data(_data_receipt) => {
+            // Data receipts themselves don't cost gas to execute, their cost is
+            // burnt at creation. What we should count, is the gas of the
+            // postponed action receipt. But looking that up would require
+            // reading the postponed receipt from the trie.
+            // Thus, the congestion control MVP does not account for data
+            // receipts or postponed receipts.
+            Ok(0)
+        }
+        ReceiptEnum::PromiseYield(_) => {
+            // The congestion control MVP does not account for yielding a promise.
+            // It does not burn any gas, so it's unclear what we could account for.
+            Ok(0)
+        }
+        ReceiptEnum::PromiseResume(_) => {
+            // The congestion control MVP does not account for resuming a promise.
+            // Similar to a data receipt, it would be difficult to find the cost
+            // of it without expensive state lookups.
+            Ok(0)
+        }
+    }
+}
+
+/// Iterate all columns in the trie holding unprocessed receipts and
+/// computes the storage consumption as well as attached gas.
+///
+/// This is an IO intensive operation! Only do it to bootstrap the
+/// `CongestionInfo`. In normal operation, this information is kept up
+/// to date and passed from chunk to chunk through chunk extra fields.
```

Comment:
I think technically it's passed through chunk headers e.g. when doing state sync we do not have chunk extra for chunks in the first block but we do have the chunk headers.